esp-modem build problems (IDFGH-7729)

IDF: 4.4.1
Verison: 0.1.17
Tool Chain: esp-2021r2-patch3-8.4.0
Target: esp32

```
Processing 3 dependencies:
[1/3] espressif/esp-modbus (1.0.3)
[2/3] espressif/esp_modem (0.1.17)
[3/3] idf (4.4.1)
```

Hi we are attempting to migrate from the older version of the esp-modem to the latest version and have come across some build issues.

#### Unknown netif type
```
In file included from /Users/ash/dev/project/managed_components/espressif__esp_modem/include/esp_modem_api.h:19,
                 from /Users/ash/dev/project/components/cellular-test/src/cellular_test.c:11:
/Users/ash/dev/project/managed_components/espressif__esp_modem/include/esp_modem_c_api_types.h:66:116: error: unknown type name 'esp_netif_t'; did you mean 'esp_err_t'?
 esp_modem_dce_t *esp_modem_new(const esp_modem_dte_config_t *dte_config, const esp_modem_dce_config_t *dce_config, esp_netif_t *netif);
                                                                                                                    ^~~~~~~~~~~
                                                                                                                    esp_err_t
```

#### Invalid cast

```
/Users/ash/dev/project/components/cellular-test/src/cellular_test.c:180:10: warning: assignment to 'esp_modem_dce_t *' {aka 'struct esp_modem_dce_wrap *'} from 'int' makes pointer from integer without a cast [-Wint-conversion]
    p_dce = esp_modem_new_dev(ESP_MODEM_DCE_BG96, &dte_config, &dce_config, p_esp_netif);
```
Where p_dce is  'esp_modem_dce_t*'

### Missing functions
```
/Users/ash/dev/project/components/cellular-test/src/cellular_test.c: In function 'cellular_connect_task':
/Users/ash/dev/project/components/cellular-test/src/cellular_test.c:170:10: error: implicit declaration of function 'esp_modem_new_dev'; did you mean 'esp_modem_reset'? [-Werror=implicit-function-declaration]
  p_dce = esp_modem_new_dev(ESP_MODEM_DCE_BG96, &dte_config, &dce_config, p_esp_netif);
          ^~~~~~~~~~~~~~~~~
          esp_modem_reset
```

My cellular-test component has `esp_modem` in the PRIV_REQUIRES list, and I can see the component has been correctly configured in the managed_components directory.

--- a/components/esp_modem/include/esp_modem_c_api_types.h
+++ b/components/esp_modem/include/esp_modem_c_api_types.h
@@ -15,6 +15,7 @@
 #pragma once
 
 #include "esp_modem_config.h"
+#include "esp_netif.h"
 
 #ifdef __cplusplus
 extern "C" {
@@ -22,7 +23,11 @@ extern "C" {
 
 typedef struct esp_modem_dce_wrap esp_modem_dce_t;
 
-struct PdpContext;
+typedef struct esp_modem_PdpContext_t {
+    size_t context_id;
+    const char *protocol_type;
+    const char *apn;
+} esp_modem_PdpContext_t;
 
 /**
  * @defgroup ESP_MODEM_C_API ESP_MODEM C API

--- a/components/esp_modem/include/generate/esp_modem_command_declare.inc
+++ b/components/esp_modem/include/generate/esp_modem_command_declare.inc
@@ -38,7 +38,7 @@
 #define BOOL_OUT(param, name) bool* _ARG(param, name)
 #define INT_OUT(param, name) int* _ARG(param, name)
 #define INTEGER_LIST_IN(param, name) const int* _ARG(param, name)
-#define STRUCT_OUT(struct_name, p1)  struct struct_name* p1
+#define STRUCT_OUT(struct_name, p1)  esp_modem_ ## struct_name ## _t* p1
 #endif
 
 #define DECLARE_ALL_COMMAND_APIS(...) \

--- a/components/esp_modem/src/esp_modem_c_api.cpp
+++ b/components/esp_modem/src/esp_modem_c_api.cpp
@@ -393,3 +393,14 @@ extern "C" esp_err_t esp_modem_reset(esp_modem_dce_t *dce_wrap)
 {
     return command_response_to_esp_err(dce_wrap->dce->reset());
 }
+
+extern "C" esp_err_t esp_modem_set_pdp_context(esp_modem_dce_t *dce_wrap, esp_modem_PdpContext_t *c_api_pdp)
+{
+    if (dce_wrap == nullptr || dce_wrap->dce == nullptr) {
+        return ESP_ERR_INVALID_ARG;
+    }
+    esp_modem::PdpContext pdp{c_api_pdp->apn};
+    pdp.context_id = c_api_pdp->context_id;
+    pdp.protocol_type = c_api_pdp->protocol_type;
+    return command_response_to_esp_err(dce_wrap->dce->set_pdp_context(pdp));
+}

